### DISCUSSION TITLE
@xstate/test network interception patterns

### DISCUSSION THREAD

**Discussion by Koleok** [statelyai/xstate#2873] [Q&A]
So far in my quest to adopt [MBT](https://en.wikipedia.org/wiki/Model-based_testing) at [filevine](https://www.filevine.com/) I have found the biggest pain point by far to be intercepting and dealing with network responses as events, distinct from user initiated events.

An open ended discussion about network interception patterns in different scenarios would be very valuable to those aspiring to adopt @xstate/test into their orgs 🙏 

A few inspirational pieces of prior art I came across are as follows, organized by test environment:
- [js-dom](https://github.com/jsdom/jsdom) / [testing-library](https://testing-library.com/) based testing
  - [John Bales](https://github.com/jtbales) [model-based-testing-tutorial](https://github.com/jtbales/model-based-testing-tutorial) which also has a [video walk through](https://www.youtube.com/playlist?list=PLdWY3IGxchhdjUKJRVyeR_-LKX_QT6mcR) 
    - I was able to draw on some of these ideas to get a nice system put together using: 
      - testing-library to render in js-dom
      - jest to mock a custom api client module
      - a global fetch mock to intercept network requests and return results from a static file
- e2e testing using [playwright](https://playwright.dev/docs/api/class-route#route-fulfill) and/or [cypress](https://www.cypress.io/)
  - [John de Stigter](https://github.com/rjdestigter)'s library [xstate-marionettist](https://github.com/rjdestigter/xstate-marionettist) which has it's own discussion here #1541 
    - John has done some sophisticated work on this to create a very high level abstraction
    - It's a bit difficult to grok his pattern of deferring and resolving network requests, but in my own efforts to use @xstate/test with playwright and cypress I have often consulted his work for these libs respectively
    - To date I have not been able to come up with a working pattern for cypress or playwright in regards to network control

**Comment by sukima** [comment]
I only ever used MBT in integration (Ember components, A.K.A. unit testing in the other Front End ecosystems) but what I do is mock the things I own. All network is handled by a network layer service and we can dependency inject a mock service when testing. This way any use of that service results in an **unfilled**/**pending** promise (A.K.A. the Deferred pattern).

In MBT this is modeled as three states: Before request (user action to trigger the network request), Loading (the promise is pending), and done (the promise has resolved). We might add an Error state as well and the MBT testing can path to it if our UI needs to handle network errors.

## Example

We use QUnit but this could be done for any framework.

```javascript
import FooBarComponent from 'project/components/foo-bar';
import FooBarPageObject from 'projects/components/foo-bar/test-helpers';
import sinon from 'sinon';
import { module, test } from 'qunit';
import { createMachine } from 'xstate';
import { createModel } from '@xstate/test';

class NetworkManager {
  #resolve;
  constructor() {
    this.reset();
  }
  finish(payload) {
    this.#resolve(payload);
    return this.promise;
  }
  reset() {
    this.promise = new Promise((resolve) => {
      this.#resolve = resolve;
    });
  }
}

function testWrapper(testFn) {
  return function (testContext, state) {
    // QUnit shows a flat list of all assertions. This allows the list to show
    // the breaks between each meta.test() as they all execute.
    testContext.assert.ok(
      true,
      `§ state ${JSON.stringify(state.value)} ← event ${JSON.stringify(state.event)}`,
    );
    return testFn(testContext, state);
  };
}

const page = new FooBarPageObject();

const testMachine = createMachine({
  id: 'foo-bar-test-machine',
  initial: 'rendered',
  states: {
    rendered: {
      meta: {
        test: testWrapper(function ({ assert }) {
          assert.ok(page.isVisible(), 'component has rendered');
        });
      },
      on: { CLICK_GO_BUTTON: 'loading' },
    },
    loading: {
      meta: {
        test: testWrapper(function ({ assert }) {
          assert.ok(page.isLoadingIndicatorVisible(), 'has a loading indicator');
        });
      },
      on: {
        FAILED_LOADING: 'errored',
        FINISH_LOADING: 'done',
      },
    },
    errored: {
      meta: {
        test: testWrapper(function ({ assert }) {
          assert.ok(page.isErrorMessageVisible(), 'shows an error message');
        });
      },
    },
    done: {
      meta: {
        test: testWrapper(function ({ assert }) {
          assert.ok(page.isSuccessMessageVisible(), 'shows a success message');
        });
      },
    },
  },
});

const testModel = createModel(testMachine).withEvents({
  async CLICK_GO_BUTTON() {
    await page.clickGo();
  },
  async FAILED_LOADING({ networkManager }) {
    await networkManager.finish({ status: 'failed' });
    networkManager.reset();
  },
  async FINISH_LOADING({ networkManager }) {
    await networkManager.finish({ status: 'ok' });
    networkManager.reset();
  },
});

const testPlans = testModel.getShortestPathPlans();

module('Integration | Components | FooBar', function (hooks) {
  hooks.beforeEach(function () {
    this.networkManager = new NetworkManager();
    // This would normally be a dependency injectable service not global fetch.
    sinon.stub(window, 'fetch').callsFake(() => this.networkManager.promise);
  });

  hooks.afterEach(function () {
    sinon.restore();
  });

  testPlans.forEach((plan) => {
    module(plan.description, function () {
      plan.paths.forEach((path) => {
        test(path.description, async function (assert) {
          let { networkManager } = this;
          let subject = new FooBarComponent();
          await subject.render();
          await path.test({ assert, networkManager });
        });
      });
    });
  });
});
```
